compiler/src/dotty/tools/dotc/transform/init/Checking.scala

```diff
@@ -0,0 +1,423 @@
+      val cls = ctor.owner
+      val classDef = cls.defTree
+      if (!classDef.isEmpty) {
+        if (ctor.isPrimaryConstructor) checkClassBody(classDef.asInstanceOf[TypeDef])
+        else checkSecondaryConstructor(ctor)
+      }
+      else if (!cls.isOneOf(Flags.EffectivelyOpenFlags))
+        ctx.warning("Inheriting non-open class may cause initialization errors", source.sourcePos)
+    }
+
+    cls.paramAccessors.foreach { acc =>
+      if (!acc.is(Flags.Method)) {
+        traceIndented(acc.show + " initialized", init)
+        state.fieldsInited += acc
+      }
+    }
+
+    tpl.parents.foreach {
+      case tree @ Block(_, parent) =>
+        val (ctor, _, _) = decomposeCall(parent)
+        checkCtor(ctor.symbol, parent.tpe, tree)
+
+      case tree @ Apply(Block(_, parent), _) =>
+        val (ctor, _, _) = decomposeCall(parent)
+        checkCtor(ctor.symbol, tree.tpe, tree)
+
+      case parent : Apply =>
+        val (ctor, _, argss) = decomposeCall(parent)
+        checkCtor(ctor.symbol, parent.tpe, parent)
+
+      case ref =>
+        val cls = ref.tpe.classSymbol.asClass
+        if (!state.parentsInited.contains(cls))
+          checkCtor(cls.primaryConstructor, ref.tpe, ref)
+    }
+
+    // check class body
+    tpl.body.foreach { checkClassBodyStat(_) }
+  }
+
+  def checkSecondaryConstructor(ctor: Symbol)(implicit state: State): Unit = traceOp("checking " + ctor.show, init) {
+    val Block(ctorCall :: stats, expr) = ctor.defTree.asInstanceOf[DefDef].rhs
+    val cls = ctor.owner.asClass
+
+    traceOp("check ctor: " + ctorCall.show, init) {
+      val ctor = ctorCall.symbol
+      if (ctor.isPrimaryConstructor)
+        checkClassBody(cls.defTree.asInstanceOf[TypeDef])
+      else
+        checkSecondaryConstructor(ctor)
+    }
+
+    (stats :+ expr).foreach { stat =>
+      val (_, effs) = Summarization.analyze(stat)(theEnv.withOwner(ctor))
+      val rebased = Effects.asSeenFrom(effs, ThisRef(state.thisClass)(null), cls, Potentials.empty)
+      rebased.foreach { check(_) }
+    }
+  }
+
+  private def nonInitError(field: Symbol)(implicit state: State) = {
+    traceIndented("initialized: " + state.fieldsInited, init)
+
+    // should issue error, use warning so that it will continue compile subprojects
+    theCtx.warning(
+      "Access non-initialized field " + field.show +
+      ". Calling trace:\n" + state.trace,
+      field.sourcePos
+    )
+  }
+
+  private def externalCallError(sym: Symbol, source: Tree)(implicit state: State) =
+    theCtx.warning(
+      "Calling the external method " + sym.show +
+      " may cause initialization errors" +
+      ". Calling trace:\n" + state.trace,
+      source.sourcePos
+    )
+
+  private def check(eff: Effect)(implicit state: State): Unit =
+    if (!state.visited.contains(eff)) traceOp("checking effect " + eff.show, init) {
+      implicit val state2: State = state.withVisited(eff)
+
+      eff match {
```

Comment:
It seems, in essence, each case here does a simple thing roughly describable in a sentence:

- `Leak` – they are bad, display a warning whenever it is detected.
- `FieldAccess` – resolve the field and issue a warning if it is not registered in `state.fieldsInited` where all initialized fields are tracked.
- `MethodCall` – check the effects of the method being called. If the method is out of the scope of the project, issue a warning.

Is my understanding correct here? If so, do you think it will make sense to add a short comment above each case? Would save a lot of time IMO for future readers.